Add events task

cecton opened the pull request:
## What These Changes Do

1. ~Explicit loop~
2. Fix a problem with ClientResponse is not a coroutine
3. DockerEvents automatically create and cancel a background task by default (breaking change)

asvetlov commented on aiodocker/docker.py:
```diff
@@ -86,8 +86,8 @@ def __init__(self,
         self.images = DockerImages(self)
         self.volumes = DockerVolumes(self)
 
-    async def close(self):
```
Please revert it back.
`session.close()` **is** a coroutine.

cecton replied:
@asvetlov sorry it's a work in progress, don't waste your time looking at it at the moment. Thanks for the remark though I was totally mistaking.

barrachri commented:
@cecton just ping when the PR is ready for the review.

cecton commented:
@barrachri ping

So here is the idea I had in #34 : 20965e0a4ba651abcee3e57ed68d1d0db64979d2

But to do that I had to add the event loop to the Docker instance: 7097cb19a9dbca10b924e2614145fb1ef86e208f

And I also installed tox, removed the need of Docker Compose and made all the test in `test_images.py` unrelated to each other (so you can run them seperately): bb24437f0708b005783eba05627628059c7bd968

I also made a lot of changes in the requirements files:
 - I added requirements.txt, a single place where you put only the library's requirements
 - I added requirements/test.txt, a single place where you put only the requirements for running the tests
 - I replaced requirements/ci.txt, it is now a single place without versions where you put the requirements for the CI (codecov)

But I have no idea why setuptools is in the requirements to be honest. I saw that the bot is updating its version but setuptools is only used in the CI itself. It doesn't make much sense to me, I would remove it if I were you.

cecton commented:
Hmmm.... I just spotted something fishy with a try...except: pass... there will be more fixes in this PR.

asvetlov commented on aiodocker/docker.py:
```diff
@@ -42,7 +42,23 @@ def __init__(self,
                  ssl_context=None,
-                 api_version='v1.26'):
+                 api_version='v1.26',
+                 loop=None):
```
Using implicit loop is preferable technique starting from python 3.5.3
aiodocker is a new library, we don't need to pass loop everywhere.

cecton replied:
I didn't know that... look what I found on the web when trying to find more information: https://groups.google.com/forum/#!topic/python-tulip/yF9C-rFpiKk

asvetlov replied:
See also https://github.com/python/asyncio/pull/452

asvetlov commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+aiohttp==2.2.3
```
I don't like this file.
If you need new requirement -- please put it into `requirements` folder.

cecton replied:
`requirements.txt` is more generally used in Python projects when you look for the requirements. I expect most Python developers to expect either a `requirements.txt` or no `requirements.txt` at all because the dependencies are in the `setup.py`.

My true goal here is only to get one place to maintain the dependencies. I can move it to `requirements/base.txt` or something if you prefer?

asvetlov replied:
Our requirements are for aiodocker developers, users install the library by `pip install aiodocker` without explicitly looking into requirement files at all.
Thus doesn't matter where we do place the file.
If we have everything in `requirements` folder -- let's put base requirements there.
`requirements/base.txt` sounds pretty good.

asvetlov commented on setup.py:
```diff
@@ -42,12 +50,7 @@
-    extras_require={
-        'test': ['pytest', 'pytest-asyncio', 'flake8'],
-    }
+    install_requires=requirements,
```
I suspect the library should not pin own requirements -- otherwise we need publish new aiodocker on every yarl/aiohttp update.

cecton replied:
Nevertheless the lib work with some versions but not all. For aiohttp I suppose we are following semantic versioning so maybe we could use `<2.3`. What do you think would be best? Simply removing any version restriction?

asvetlov replied:
Relax it to non-strict check, e.g. `aiohttp>=2.0` where `2.0` is minimal supported version.
Sorry, I don't follow what minimal aiohttp version is supported by aiodocker.

cecton replied:
I will figure that out easily with tox

cecton replied:
2.0 works, 1.3.5 failed

asvetlov commented:
Yes, setuptools is not needed in requirements

asvetlov commented:
Other thing: I think better to split new feature (a task for events monitoring) and improvements of existing infrastructure.

cecton commented on tests/test_images.py:
```diff
@@ -1,21 +1,25 @@
 from aiodocker.exceptions import DockerError
 
 
+def _random_name():
```
Thx ^_^

cecton commented on tests/test_images.py:
```diff
@@ -1,21 +1,25 @@
               "aiodocker/master/tests/docker/Dockerfile")
 
-    tag = "image:1.0"
+    tag = "%s:1.0" % _random_name()
```
I checked that in the past and apparently the devs at Python don't want to drop the `%` syntax because is handy and small to use. Plus they even added a new one:

```python
you = "barrachri"
print(f"Hello {you}")
```

So.... well I disagree but I really don't mind at all changing it for consistency sake.

barrachri replied:
As I said it just a matter of consistency :) remember that is `.format()` and not `f""`, that is available only from python 3.6.

cecton replied:
Yes but `%` is not going to be deprecated and is available ^_^ but anyway...

cecton commented on tests/test_images.py:
```diff
@@ -1,21 +1,25 @@
-    async for output in stream:
-        if "Successfully tagged image:1.0\n" in output:
-            break
+    async for _ in stream:  # noqa: F841
```
That's the funny part of this PR. If you don't loop, you're image will not finish to build and the code will break. Maybe that only happens if the `stream=True` is active though. I know it's the case for docker-py and I just checked with a big image with aiodocker and it behaves the same.

cecton replied:
Now the funny part is that the previous code has problems too. I first tried to use it and catch from the output and assert on the output. But I realized that the "output" is no string, it's a dict, and my assert failed. The actual log line is in `output['stream']`.

So I updated the code and then the test failed because depending on the version of Docker, you may get an upper case, a lower case here and there, etc...

So in the end I felt like the only interesting thing to do here is to wait for the build to complete because this for loop doesn't actually do anything else than that.

cecton replied:
So actually we can remove the for loop with the `stream=True` and it will behave the same.

cecton commented on aiodocker/jsonstream.py:
```diff
@@ -42,7 +42,7 @@ def __aiter__(self):
 
         # response error , it has been closed
-        await self._response.close()
+        self._response.close()
```
No I actually fixed it. This is no coroutine apparently and the try...except...pass on the top level was hiding the exception. Commit message: 1d3abd8c570eb66e2fcc9c0b297c9d2a9153c7a3

cecton commented:
@asvetlov @barrachri okay, ready for review.

cecton commented on aiodocker/docker.py:
```diff
@@ -500,20 +503,25 @@ def _transform_event(self, data):
-            try:
-                await self.json_stream.close()
-            except:
-                pass
```
@barrachri the error in the `close()` used to be hidden by this try...except: pass

barrachri replied:
Ok, it seems harmless, I'll check this `.close()` in the future PRs.

barrachri commented:
@cecton I plan to close the review tomorrow, I didn't find time during the weekend :/

cecton commented:
@barrachri no problem for me! :) Thanks.

To be honest, at this point I'm more evaluating aiodocker to see if I can replace my own one. I'm still very dubious about it especially since I discovered how I can properly override the methods in docker-py.

asvetlov commented:
@barrachri please feel free to merge if the PR looks good to you.

BTW we can raise election of @cecton to aiodocker team.
She is a member of aiohttp already, why not?

barrachri commented:
@cecton I re-run the tests after the last merges, there seems to be some errors.

Can you have a look?

cecton commented:
@barrachri okay I just checked and close() on `_JsonStreamResult` has been renamed to `_close`. Should I simply use that one?

barrachri commented:
Yes, thanks!

cecton commented:
@barrachri done

barrachri commented:
Thanks @cecton !
